src/main/java/org/ow2/proactive/connector/iaas/cloud/provider/jclouds/JCloudsProvider.java

```diff
@@ -216,14 +217,30 @@ protected String buildScriptToExecuteString(InstanceScript instanceScript) {
     }
 
     private RunScriptOptions buildScriptOptions(InstanceScript instanceScript) {
-        logger.info("Script options: userLogin=" + getVmUserLogin());
-        return Optional.ofNullable(instanceScript.getCredentials())
-                       .map(credentials -> RunScriptOptions.Builder.runAsRoot(false)
-                                                                   .overrideLoginCredentials(new LoginCredentials.Builder().user(credentials.getUsername())
-                                                                                                                           .password(credentials.getPassword())
-                                                                                                                           .authenticateSudo(false)
-                                                                                                                           .build()))
-                       .orElse(RunScriptOptions.Builder.overrideLoginUser(getVmUserLogin()));
+        return Optional.ofNullable(instanceScript.getCredentials()).map(credentials -> {
+            if (credentials.getPrivateKey() != null) {
+                String username = Optional.of(credentials.getUsername())
+                                          .filter(StringUtils::isNotEmpty)
+                                          .filter(StringUtils::isNotBlank)
+                                          .orElse(getVmUserLogin());
+                logger.info("Key based authentication is provided. Script is going to be executed with the given private key and username=" +
+                            username);
+                return RunScriptOptions.Builder.runAsRoot(false)
+                                               .overrideLoginUser(username)
+                                               .overrideLoginPrivateKey(credentials.getPrivateKey());
+            } else {
+                logger.info("Credentials are provided. Script is going to be executed with the given password and username=" +
```

Comment:
I think you should double check if the username:password creds are actually provided

Reply:
That was the way it was implemented before this commit. I agree let's improve this with a check.